File: src/main/scala/org/foolish/fvm/scubc/ValueFiroe.scala
package org.foolish.fvm.scubc

import org.foolish.ast.AST

/**
 * ValueFiroe represents a finalized value in the computation.
 * For now, supports integral long values.
 */
case class ValueFiroe(override val ast: AST, value: Long) extends FiroeWithoutBraneMind(ast):

  def this(value: Long) = this(null, value)

  /** Returns the integral value stored in this ValueFiroe */
  override def getValue: Long = value

  /** ValueFiroe is never abstract as it represents a concrete value */
  def isAbstract: Boolean = false

  override def toString: String = value.toString

  override def clone: FIR = this // immutable constant

File: foolish-core-java/src/main/java/org/foolish/fvm/ubc/ValueFiroe.java
package org.foolish.fvm.ubc;

import org.foolish.ast.AST;

/**
 * ValueFiroe represents a finalized value in the computation.
 * For now, supports integral long values.
 */
public class ValueFiroe extends FiroeWithoutBraneMind {
    private final long value;

    public ValueFiroe(AST ast, long value) {
        super(ast);
        this.value = value;
    }

    public ValueFiroe(long value) {
        this(null, value);
    }

    /**
     * Returns the integral value stored in this ValueFiroe.
     */
    public long getValue() {
        return value;
    }

    /**
     * ValueFiroe is never abstract as it represents a concrete value.
     */

    public String toString() {
        return String.valueOf(value);
    }

    public FIR clone(){
        return this; // I'm a constant
    }
}
